crates/hir/src/source_analyzer.rs

```diff
@@ -459,24 +459,68 @@ pub(crate) fn resolve_hir_path(
     resolve_hir_path_(db, resolver, path, false)
 }
 
+fn resolve_assoc_type(
+    db: &dyn HirDatabase,
+    ty: TypeNs,
+    unresolved: PathSegment,
+) -> Option<PathResolution> {
+    let traits_from_env = match ty {
+        TypeNs::GenericParam(param_id) => {
+            let predicates = db.generic_predicates_for_param(param_id);
+            let mut traits: Vec<_> = predicates
+                .iter()
+                .filter_map(|pred| match &pred.value.value {
+                    WhereClause::Implemented(tr) => Some(tr.hir_trait_id()),
+                    _ => None,
+                })
+                .collect();
+            // Handle `Self::Type` referring to own associated type in trait definitions
+            if let GenericDefId::TraitId(trait_id) = param_id.parent {
+                traits.push(trait_id);
+            }
+
+            traits
+        }
+        TypeNs::SelfType(impl_id) => match db.impl_trait(impl_id) {
+            None => vec![],
+            Some(trait_ref) => vec![trait_ref.value.hir_trait_id()],
+        },
+        _ => return None,
+    };
+    // FIXME supertraits
```

Comment:
We got a function for this in `hir_ty`, I guess it would make sense to move it to `hir_def`?(I think it's implementation solely relies on `hir_def`). I have no clue in which module that would belong in that case though, unless a new `utils` module like it is in `hir_ty`
https://github.com/rust-analyzer/rust-analyzer/blob/25201b2dad7b4b0d41494e238ebf643ad7ad8cd6/crates/hir_ty/src/utils.rs#L81

Reply:
I probably want to get rid of `all_super_traits`, but it might be nice to turn `all_super_trait_refs` into a query.